### ISSUE TITLE
Size OpenSearch Domain based on Expected traffic

### ISSUE THREAD

**Issue by chelma** [arkime/aws-aio#56]
## Description
Currently, the size of the OpenSearch Domain (number of nodes, instance type) is static and not configurable by the user.  We should enable the user to indicate an expected traffic volume/data retention policy and provision OpenSearch capacity according to the user's specification. 

Considerations:
* @awick provided the following equations based on his experience.  Most interestingly is the metadata sizing factor (0.03) which is used as a constant relative to the total raw packet volume.
```
SessionsGB = ($replicas + 1) * ($spiDays * 24 * 60 * 60) * $ExpectedGbps/8 * 0.03
NodesByGB = $SessionsGB/$GBPerOSNode
Nodes = MAX($NodesByGB, $replicas + 1)
```
* The OpenSearch sizing docs [1] introduce another interesting facet, which is that there are three significant factors impacting how much of the EBS capacity is actually usable - indexing overhead (10%), linux system reserved space (5%), and OpenSearch Service overhead (20%).  As a result, the docs recommend scaling your minimum EBS capacity by a constant factor of 1.45.

[1] https://docs.aws.amazon.com/opensearch-service/latest/developerguide/sizing-domains.html

Note: This is a sub-task of https://github.com/arkime/cloud-demo/issues/34

## Acceptance Criteria
* The user should be able to specify an expected traffic volume when calling `create-cluster`
* Demonstrate the ability of the CLI to set and update the OpenSearch Domain size.

**Comment by awick** [comment]
Those values are already included in the "magic formula"

**Comment by chelma** [comment]
@awick Ah - thanks for letting me know.  We would have been seriously over-provisioning.

**Comment by chelma** [comment]
Here's a couple scenarios for how much storage we'll need based on the traffic amount.  The minimum/maximum traffic were figures established in (https://github.com/arkime/aws-aio/issues/52), and represent the range of traffic the user can specifically configure as expected.  The minimum traffic amount represents a single capture node instance, and the max represents the most that can sent over a single GWLB endpoint.

```
MIN TRAFFIC, DEFAULT OPTIONS:
Scenario: 2 Gibps traffic, 30-day SPI retention, 1 replica
SPI data to store = 2 copies * 2592000 retention seconds * 0.25 GB/sec * 0.03 = 38,880 GB of storage

MAX TRAFFIC, DEFAULT OPTIONS:
Scenario: 100 Gibps traffic, 30-day SPI retention, 1 replica
SPI data to store = 2 copies * 2592000 retention seconds * 25 GB/sec * 0.03 = 3,888,000 GB of storage
```

The max-traffic scenario actually exceeds the maximum allowable storage for a single OpenSearch domain of 3 PiB [1].  This brings up a few considerations.

* 2 Gibps is a lot of storage, and using it as our quanta for capture capacity is probably not right.  Moving to a smaller amount will give us much more flexibility in matching the user's traffic volume to a reasonable amount of cloud capacity.  Taking a brief look, m5.large is half the capacity and price of our current capture node choice, m5.xlarge.  We could easily bump down to the smaller instance type and make our quanta 1 Gibps.
* By default, OpenSearch Service hands out the `r5.*` family of instances for master and data nodes, so we'll probably start with those ourselves.  `r5.large` is the smallest instance of that family and can handle EBS volumes from 1-2 TB.  `r5.4xlarge` will handle 6-12 TB, `r5.12xlarge` will handle 12-24 TB.
* We will want a way to provision a "toy" Domain that doesn't assume completely maxed out traffic for the specified capture capacity.  Alternatively, we could give users more levers to configure the OpenSearch Domains.
* We will need a way to inform the user they've exceeded the storage limit of a single Domain

[1] https://docs.aws.amazon.com/opensearch-service/latest/developerguide/limits.html

**Comment by awick** [comment]
Yes we should be using what is passed in for the OS sizing, and not basing it on what capture COULD generate. If possible we should also allow folks to pass in Gbps floats. I agree there will be some tuning we need to do, especially at the low and high bans, and right now we are already over estimating (with the 0.03). This will be another opportunity for learning.

**Comment by chelma** [comment]
Per convo w/ @awick, it sounds like we have alignment.

* Drop the capture node size to `m5.large`
* Accept floats for the expected traffic.  Quantize this for capture capacity and respect the real value for OS Domain sizing.
* Update our "minimum" traffic in the code to a much smaller number so users don't end up with like 20 OpenSearch Domain nodes just for turning this thing on with default values.
* Have a "toy-tier" data node instance type we can use for the default/minimal configuration.

**Comment by chelma** [comment]
Ah - made a mistake above.  100 Gbps is 12.5 GBps, not 25 GBps.  That means the max storage example would be 1,944,000 GB.  Still a lot, but not over the max for a single domain.

`t3.small` will hold 10-100 GB and is 1/6 the price of a `r5.large`; `t3.medium` will hold 10-200 GB and is 1/3 the price of a `r5.large`.  Unclear on the performance of these two nodes for write-heavy use cases, we'll have to experiment.  But one of them might be a good pick for our "toy tier" data node.  Inclined to go with the smaller one until we hear otherwise.

Reversing our calculation, we get something like the following to predict the traffic a given amount of storage can handle:

```
Gbps traffic per GB storage = 8 / ($replicas + 1) / ($spiDays * 24 * 60 * 60) / 0.03

8 / (1 + 1) / (30 * 24 * 60 * 60) / 0.03 => 5.144 x 10 ^ (-5) Gbps traffic / GB storage
```

**Comment by chelma** [comment]
Some proposed heuristics for our data node selection algorithm:
* We should always have at least two data nodes
* To simplify our lives, we'll just max out the storage for a given instance type if it has a range.
* We'll calculate our instance type spread using the default retention and replica options as our starting point.
* We can apparently choose between gp2 and gp3 EBS volume types.  It's unclear what offers the better overall value, but gp3 is apparently more complex to configure so we'll use gp2 to start off with.
* The gp2 EBS volumes pricing scales linearly ($0.14/TB-hr).  The cost ratio between the instance and the storage for the instance increases as the instance size increases (large instances cost more for the same storage).  Additionally, having more, smaller instances provides better resiliance and decreases the cost delta for adding another node of storage.  Therefore, there seems to be an incentive to have more smaller instances than fewer large ones.
* There's a maximum of 10 t2/t3 instances per domain, and a default limit of 80 instances for a domain total (expandable up to 200 instances).
* It's implied that dedicated master nodes don't count against the instance limit

Therefore, we end up with something like:

```
 Traffic      | Storage   | Instance Type  | Storage/Node | Published Price/hr (us-east-2)
-----------------------------------------------------------------------------
0-0.5 Gbps    | 0-1 TB    | t3.small       | 100 GB       | $0.0208
0.5-4 Gbps    | 1-80 TB   | r5.large       | 1 TB         | $0.126
4-24.6 Gbps   | 80-480 TB | r5.4xlarge     | 6 TB         | $1.008
24.6+ Gbps    | 480-960 TB| r5.12xlarge    | 12 TB        | $3.024
```

If a user needs more than that maximum, they'll need to manually request an instance limit increase and we can just keep adding more of the largest instance type.

**Comment by chelma** [comment]
Looking at the suggestions for dedicated master node instance types [1], it seems there's a both a recommendation based on instance type and number of shards, with shards being a function of total data stored.  The smallest instance type recommended (m5.large) can handle 10k shards, with each shard being around 40 GB [2], for a total of 400 TB of data (before replication), and the next size up able to handle 1.2 PB.  Therefore, the driving factor is the number of instances in the domain.  We can simply do this scheme to fit the dedicated master nodes:

* Always have 3, per recommendations
* If pre-replication data amount over threshold, use larger size
* Otherwise, base it on the number of data instances 


[1] https://docs.aws.amazon.com/opensearch-service/latest/developerguide/managedomains-dedicatedmasternodes.html
[2] https://docs.aws.amazon.com/opensearch-service/latest/developerguide/sizing-domains.html#bp-sharding

**Comment by chelma** [comment]
A couple comments.

* The pricing in the table above is for the EC2 variant of these instances.  OpenSearch runs its own instance variants with their own pricing [1].
* We should be using the Graviton instances (e.g. r6g.*) for the non-toy instance types.  They provide better prices for the same EBS storage capacity.

[1] https://aws.amazon.com/opensearch-service/pricing/

**Comment by chelma** [comment]
Finally got to the point of trying this stuff out for real in a cluster, got an interesting error from CloudFormation on OS Domain spinnup:

```
Resource handler returned message: "Invalid request provided: Cannot use graviton instances with non-graviton instances. Please use graviton instances as data nodes or use non-graviton instances as master nodes. (Service: OpenSearch, Status Code: 400, Request ID: 4ae4cee6-c8fd-465b-aaba-58caa0ddf048)" (RequestToken: 2579f6f0-36ea-d88c-41bf-7cddcd5b9652, HandlerErrorCode: InvalidRequest)
```

**Comment by chelma** [comment]
Ugh, of course, there are no "toy"-sized Graviton Data Node instance types.  So we'll need to have an additional check for really small clusters to switch their Master Node instance type to non-gGaviton.

**Comment by chelma** [comment]
Double-ugh! Plus a YUCK!  New, delightful edge case to deal with:

```
Invalid request provided: You must choose an even number of data nodes for a two Availability Zone deployment (Service: OpenSearch, Status Code: 400, Request ID: 618481cf-7273-4fe5-b41c-95ff50fff155)
```
